pkg/storage/replicate_queue.go

```diff
@@ -53,6 +53,15 @@ const (
 	// in high latency clusters, and not allowing enough of a cushion can
 	// make rebalance thrashing more likely (#17879).
 	newReplicaGracePeriod = 5 * time.Minute
+
+	// addToPurgatoryReplicaCountThreshold is the count of replicas
+	// beyond which replicas that cannot be up-replicated are sent to
+	// purgatory for eager up-replication. Eager up-replication is
+	// mostly a concern when cluster sizes are small. If a cluster with
+	// many ranges is unable to allocate targets for up-replication, the
+	// purgatory will be filled with replicas and end up doing a lot of
+	// useless busy work. This is especially true for single-node clusters.
+	addToPurgatoryReplicaCountThreshold = 1000
```

Comment:
Ignore my other comment as this addresses my concern. 1000 seems a bit arbitrary. Any insight into how you chose that value? Why not 100? Should probably have something to do with the overhead of placing these replicas back in purgatory when a spurious gossip occurs, vs how long it takes to process these replicas when a new store is added.